rust/connlib/tunnel/src/control_protocol/client.rs

```diff
@@ -17,40 +17,34 @@ use webrtc::{
     },
 };
 
-use crate::{ClientState, ControlSignal, Error, PeerConfig, Request, Result, Tunnel};
+use crate::{peer::Peer, ClientState, Error, PeerConfig, Request, Result, Tunnel};
 
 #[tracing::instrument(level = "trace", skip(tunnel))]
-fn handle_connection_state_update<C, CB>(
-    tunnel: &Arc<Tunnel<C, CB, ClientState>>,
+fn handle_connection_state_update<CB>(
+    tunnel: &Arc<Tunnel<CB, ClientState>>,
     state: RTCPeerConnectionState,
     gateway_id: GatewayId,
     resource_id: ResourceId,
 ) where
-    C: ControlSignal + Clone + Send + Sync + 'static,
     CB: Callbacks + 'static,
 {
     tracing::trace!("peer_state");
     if state == RTCPeerConnectionState::Failed {
-        tunnel
-            .awaiting_connection
-            .lock()
-            .remove(&resource_id.into());
+        let mut role_state = tunnel.role_state.lock();
+        role_state.awaiting_connection.remove(&resource_id);
+        role_state.gateway_awaiting_connection.remove(&gateway_id);
+
```

Comment:
Can we scope this? (or explicitly drop the `MutexGuard`)

Not for performance reasons but to reduce the chances of a deadlock.

Reply:
I am planning to make these fields private and introduce functions on `ClientState` instead!~